### PR TITLE
default apiDiscovery in SDK

### PR DESCRIPTION
**Pull request by invincibleJai**
**Fixes:**

https://issues.redhat.com/browse/HAC-390

**Description:**

- [x] Makes the prop for apiDiscovery optional in AppInitSDK
- [x] Uses the provided fetch call, make a call to get the resources it used
- [x] Setup the SDK store with the information provided

**Test Setup:**
Don't need to pass `apiDiscovery ` in `AppInitSDK`

https://github.com/openshift/console/blob/master/frontend/public/components/app.jsx#L493

**Tests**
- [ ] Add/update tests

### CODE REVIEW

File: frontend/packages/console-dynamic-plugin-sdk/src/app/k8s/api-discovery/api-discovery.ts
Comment (invincibleJai) [inline frontend/packages/console-dynamic-plugin-sdk/src/app/k8s/api-discovery/api-discovery.ts:146]: `receivedResources` is in SDK fo dispatched to that as it updates all resources, can create new selectors and break it down to only have models and this will update only `k8s.resources.models` in store. 

But then testing default API discovery will not be easy as console needs other resources as well to be set for testing the changes

File: frontend/packages/console-dynamic-plugin-sdk/src/app/k8s/api-discovery/api-discovery.ts
Comment (invincibleJai) [inline frontend/packages/console-dynamic-plugin-sdk/src/app/k8s/api-discovery/api-discovery.ts:133]: we can only store models and skip others but we get all these resources with the same calls so we can have it too.

**Comment by invincibleJai** [comment]
/cc @andrewballantyne @christianvogt 

PTAL if you get time, thanks.

File: frontend/packages/console-dynamic-plugin-sdk/@types/sdk/index.d.ts
Comment (florkbr) [inline frontend/packages/console-dynamic-plugin-sdk/@types/sdk/index.d.ts:4]: @invincibleJai just wanted to strike up the same debate we have in the past: do we want to track the console version explicitly (since that is console specific? what does that mean for the HAC UI?). I see that this is used below to essentially help invalidate cached resources, so perhaps we simply rename/remap this to be a bit more generic?

File: frontend/packages/console-dynamic-plugin-sdk/src/app/__tests__/AppInitSDK.spec.tsx
Comment (florkbr) [inline frontend/packages/console-dynamic-plugin-sdk/src/app/__tests__/AppInitSDK.spec.tsx:95]: 👍 for tests!

File: frontend/packages/console-dynamic-plugin-sdk/src/app/__tests__/AppInitSDK.spec.tsx
Comment (florkbr) [inline frontend/packages/console-dynamic-plugin-sdk/src/app/__tests__/AppInitSDK.spec.tsx:92]: should we spy on initApiDiscoverySpy again here and ensure it is not called when a store instance is provided to cover all bases?

File: frontend/packages/console-dynamic-plugin-sdk/src/app/k8s/api-discovery/discovery-cache.ts
Comment (florkbr) [inline frontend/packages/console-dynamic-plugin-sdk/src/app/k8s/api-discovery/discovery-cache.ts:43]: I know this was migrated from the console so not a blocker - but it seems odd to me to resolve a promise with null. I feel like this function should return the cached resources and in all other cases reject the promise with an error for clarity purposes (IE if the cached resources are out of date). 

Curious on your thoughts and not blocking at all as this will require additional logic changes.

File: frontend/packages/console-dynamic-plugin-sdk/src/app/k8s/api-discovery/discovery-cache.ts
Comment (florkbr) [inline frontend/packages/console-dynamic-plugin-sdk/src/app/k8s/api-discovery/discovery-cache.ts:5]: Curious - is there any difference between this function and where we got it from the console aside from the local storage key? https://github.com/openshift/console/blob/master/frontend/public/module/k8s/get-resources.ts#L34

Wondering if we could make that a parameter to the function and de-dupe unless we want to explicitly keep this private in both locations.

File: frontend/packages/console-dynamic-plugin-sdk/src/app/k8s/api-discovery/api-discovery.ts
Comment (florkbr) [reply frontend/packages/console-dynamic-plugin-sdk/src/app/k8s/api-discovery/api-discovery.ts:133]: @invincibleJai so you're saying we don't need the other resources just yet/don't store them, but you've kept them in case we do in the future?

File: frontend/packages/console-dynamic-plugin-sdk/src/app/k8s/api-discovery/api-discovery.ts
Comment (florkbr) [inline frontend/packages/console-dynamic-plugin-sdk/src/app/k8s/api-discovery/api-discovery.ts:161]: Nit: can we store `60000` in a const at the top of the file so we can find it easier later?
`const API_DISCOVERY_POLL_INTERVAL = 60000;`

File: frontend/packages/console-dynamic-plugin-sdk/src/app/k8s/api-discovery/api-discovery.ts
Comment (florkbr) [inline frontend/packages/console-dynamic-plugin-sdk/src/app/k8s/api-discovery/api-discovery.ts:152]: Nit/confusion: Did we need to rename from `startAPIDiscovery` to `startAPIDiscoverySDK`? If we rename I was thinking add `Poll` to the end to indicate it's a polling function.

File: frontend/packages/console-dynamic-plugin-sdk/src/app/k8s/api-discovery/api-discovery.ts
Comment (florkbr) [reply frontend/packages/console-dynamic-plugin-sdk/src/app/k8s/api-discovery/api-discovery.ts:133]: Oh. I think I see what you are getting at. If we don't pass all this along, the `receivedResources` action will break?

File: frontend/packages/console-dynamic-plugin-sdk/src/app/k8s/api-discovery/api-discovery.ts
Comment (invincibleJai) [reply frontend/packages/console-dynamic-plugin-sdk/src/app/k8s/api-discovery/api-discovery.ts:152]: will keep it `startAPIDiscovery`

File: frontend/packages/console-dynamic-plugin-sdk/src/app/k8s/api-discovery/discovery-cache.ts
Comment (invincibleJai) [reply frontend/packages/console-dynamic-plugin-sdk/src/app/k8s/api-discovery/discovery-cache.ts:5]: my intention was to keep this private as it does not make sense to expose it from SDK.

File: frontend/packages/console-dynamic-plugin-sdk/src/app/k8s/api-discovery/api-discovery.ts
Comment (invincibleJai) [reply frontend/packages/console-dynamic-plugin-sdk/src/app/k8s/api-discovery/api-discovery.ts:133]: yeah it will,  i used exsiting selector/reducers what we have.

File: frontend/packages/console-dynamic-plugin-sdk/@types/sdk/index.d.ts
Comment (invincibleJai) [reply frontend/packages/console-dynamic-plugin-sdk/@types/sdk/index.d.ts:4]: yeah nice point, currently `window. SERVER_FLAGS. consoleVersion` is being set by OCP backend [0] and the explicit cache invalidation logic is based on it and sure it won't make sense to have it in SDK as HAC no way will be having it. So it does not make sense to have it here. 

https://github.com/openshift/console/blob/master/pkg/server/server.go#L656

Have removed it and not perform explicit cache validation based on consoleVersion and it does that auto for now i.e it does update the redux with resources in LS, remove the  LS key and make the call to openApi again and thus new response would be cached

File: frontend/packages/console-dynamic-plugin-sdk/src/app/k8s/api-discovery/discovery-cache.ts
Comment (invincibleJai) [reply frontend/packages/console-dynamic-plugin-sdk/src/app/k8s/api-discovery/discovery-cache.ts:43]: make sense, updated it

File: frontend/packages/console-dynamic-plugin-sdk/src/app/__tests__/AppInitSDK.spec.tsx
Comment (invincibleJai) [reply frontend/packages/console-dynamic-plugin-sdk/src/app/__tests__/AppInitSDK.spec.tsx:94]: updated.

File: frontend/packages/console-dynamic-plugin-sdk/@types/sdk/index.d.ts
Comment (florkbr) [reply frontend/packages/console-dynamic-plugin-sdk/@types/sdk/index.d.ts]: @invincibleJai do we know without the flag how many times we'll make the request to update the resource? I think doing it auto should be ok for now, but I'm assuming we added the check for the console version to avoid making too many calls.

**Review by florkbr — APPROVED** [review]

File: frontend/packages/console-dynamic-plugin-sdk/@types/sdk/index.d.ts
Comment (invincibleJai) [reply frontend/packages/console-dynamic-plugin-sdk/@types/sdk/index.d.ts]: the number of calls are unchanged, as we not invalidating cache so will return with cached data and in both cases, `startAPIDiscovery` is being called

https://github.com/openshift/console/pull/10903/files#diff-0403e41025b34d5ee8caf5b6c27bc0b642a8c52e567111e015a5e28430a5304dR171-R177

The console version does not come from openApi call but the OCP backend sets it and SDK should be agnostic of it as HAC backend may not provide that.

File: frontend/packages/console-dynamic-plugin-sdk/src/app/k8s/api-discovery/api-discovery.ts
Comment (andrewballantyne) [inline frontend/packages/console-dynamic-plugin-sdk/src/app/k8s/api-discovery/api-discovery.ts:71]:
```suggestion
  const apiResourceData = await consoleFetchJSON(`${k8sBasePath}/apis`);
```

**Review by andrewballantyne — CHANGES_REQUESTED** [review]
A few comments, sorry for the delay in my review.

File: frontend/packages/console-dynamic-plugin-sdk/src/app/k8s/api-discovery/discovery-cache.ts
Comment (invincibleJai) [reply frontend/packages/console-dynamic-plugin-sdk/src/app/k8s/api-discovery/discovery-cache.ts]: updated it to async, no specific reason as this code was from the console so didn't modified then but not it's updated.

File: frontend/packages/console-dynamic-plugin-sdk/src/app/k8s/api-discovery/discovery-cache.ts
Comment (invincibleJai) [reply frontend/packages/console-dynamic-plugin-sdk/src/app/k8s/api-discovery/discovery-cache.ts]: updated

File: frontend/packages/console-dynamic-plugin-sdk/src/app/AppInitSDK.tsx
Comment (invincibleJai) [reply frontend/packages/console-dynamic-plugin-sdk/src/app/AppInitSDK.tsx]: Thanks , updated

**Comment by invincibleJai** [comment]
> A few comments, sorry for the delay in my review.

Thanks @andrewballantyne, have made the changes based on input. PTAL when you get time cc @florkbr

**Review by andrewballantyne — APPROVED** [review]
Awesome, this is a great change -- but not one that OpenShift Console will see. This is a good step forward for the SDK though! 

No docs to be had here, everything is documented through the Core SDK.
/label px-approved
/label docs-approved

This will not be testable by QE. The change is a prop at the top level to optional, but OpenShift Console has its own implementation and will always pass it. Everything else is additional code under this so no way to test this by standing it up on a cluster. I have tested it locally by modifying the calling apiDiscovery code so it invokes the new logic. Everything looks good and works -- HAC-Core will be the official test of this functionality as we won't have a custom implementation of the apiDiscovery.
/label qe-approved

/hold
@florkbr Any last review items before we unhold & merge?

**Comment by florkbr** [comment]
Everyhting looks good to me! Thanks @invincibleJai and @andrewballantyne 👍 

/hold remove

**Pull request merged**
